### PR TITLE
[AIRFLOW-5659] - Add support for ephemeral storage on KubernetesPodOp…

### PR DESCRIPTION
**Pull request by leonardoam**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-5659) issues and references them in the PR title.

### Description

- [X] Add support for specifying requests and limits on ephemeral-storage for KubernetesPodOperator.

### Tests

- [X] My PR adds check on unit tests: test_gen_pod and test_gen_pod_extract_xcom

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by OmerJog** [comment]
@leonardoam can you rebase ?

**Comment by leonardoam** [comment]
Hi @OmerJog, I've rebased but some checks seems to have failed. I'll check what happened later.

**Comment by konpap94** [comment]
this would be an incredibly helpful addition

**Comment by leonardoam** [comment]
I've rebased again.
It seems Travis is broken or taking too long to run.
I ran tests locally using breeze and it's fine. The branch seems fine to merge but I accept some more review of how to make tests pass.

File: airflow/kubernetes/pod.py [hunk 1]
```diff
@@ -33,45 +33,66 @@ class Resources(K8SModel):
...
+            self.limit_memory is not None or \
+            self.limit_gpu is not None or \
+            self.limit_ephemeral_storage is not None
 
     def has_requests(self):
         """Whether resource has requests"""
-        return self.request_cpu is not None or self.request_memory is not None
+        return self.request_cpu is not None or \
+            self.request_memory is not None or \
+            self.is_empty_resource_request is not None
```
Comment (konpap94) [inline airflow/kubernetes/pod.py]: I think this line is wrong. I believe you're trying to reference `self.request_ephemeral_storage` and not `self.is_empty_resource_request` which would resolve always to True

Comment (leonardoam) [reply airflow/kubernetes/pod.py]: Thanks @konpap94 , will fix this

**Comment by ramkrishnan8994** [comment]
@leonardoam @konpap94  - When this will be merged. Causing issues for us and this might help us.

**Comment by potiuk** [comment]
@leonardoam @konpap94 -> We had some intermittent problems with Travis. They are now solved so you can rebase and there should be no more problems caused by environment. Apologies for the problems. We've suffered from that quite a lot but it's fixed now!

I am happy to review it after it is rebased/green

**Comment by leonardoam** [comment]
Nice, thanks @potiuk ! I will proceed rebasing.

**Comment by leonardoam** [comment]
Done

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Waiting for the Travis job to complete

**Comment by leonardoam** [comment]
An integration test is failing, I'll try to fix it asap.

**Comment by potiuk** [comment]
Please rebase to latest master @leonardoam - we had a six drama on Travis but it's solved now.

**Comment by potiuk** [comment]
Please rebase to latest master @leonardoam - we had a six drama on Travis but it's solved now.

**Comment by leonardoam** [comment]
@potiuk branch rebased. Thx for your help!

**Comment by konpap94** [comment]
Is there any update on this being merged in?

**Comment by yustoris** [comment]
Is any progress?

**Comment by leonardoam** [comment]
Changes are ready to be merged. Waiting for one of the commiters to merge into master.

**Review by kaxil — APPROVED** [review]

**Comment by konpap94** [comment]
🙏

**Comment by ramkrishnan8994** [comment]
When will this be added to apache-airflow 1.10.10 pip package. Need this ephemeral storage support in one of the projects.

**Comment by kaxil** [comment]
> When will this be added to apache-airflow 1.10.10 pip package. Need this ephemeral storage support in one of the projects.

In 1.10.11

**Comment by ramkrishnan8994** [comment]
> > When will this be added to apache-airflow 1.10.10 pip package. Need this ephemeral storage support in one of the projects.
> 
> In 1.10.11

If I had to integrate this change into my pip setup, how would you suggest i do it. I have 1.10.9 running now

**Comment by ramkrishnan8994** [comment]
> > When will this be added to apache-airflow 1.10.10 pip package. Need this ephemeral storage support in one of the projects.
> 
> In 1.10.11

When is this release planned?

**Comment by kaxil** [comment]
> > > When will this be added to apache-airflow 1.10.10 pip package. Need this ephemeral storage support in one of the projects.
> > 
> > 
> > In 1.10.11
> 
> When is this release planned?

In 1-2 months

**Comment by ramkrishnan8994** [comment]
> > > > When will this be added to apache-airflow 1.10.10 pip package. Need this ephemeral storage support in one of the projects.
> > > 
> > > 
> > > In 1.10.11
> > 
> > 
> > When is this release planned?
> 
> In 1-2 months

If I had to integrate this change into my pip setup, how would you suggest i do it. I have 1.10.9 running now

**Comment by dispensable** [comment]
The default value is "0" if we haven't set ephemeral-storage and even our cluster have no GPU node there is still a "nvidia.com/gpu" limit with "0" ? Those default values are unreasonable.

**Comment by leonardoam** [comment]
Hi @dispensable, are you having trouble with it? Could you explain better what is happening?

I checked the code and it seems that if you left resource values unset, they are left unset in kubernetes.

I've checked on our environment and describing a pod that only sets cpu and memory resources using kubectl gave me the following output:

```
    Limits:
      cpu:     500m
      memory:  1Gi
    Requests:
      cpu:     100m
      memory:  300Mi
```

Other resources does not appear in it.

If you have a different behavior, let me know. Could you tell me what Airflow version you are using?

**Comment by dispensable** [comment]
hi @leonardoam , we are using airflow 1.10.11. When I migrate from 1.10.11 our dags with KubernetesPodOperator are not working cause their `ephemeral-storage` got a "0" value in `kubectl get pod -o yaml`. So the pod will be evicted immediately after launch. All dags settled resource field before migration are effected. We are using v1.18.4 k8s infra.

**Comment by leonardoam** [comment]
HI @dispensable , I think this is related to this two issues:

https://github.com/apache/airflow/issues/9812
https://github.com/apache/airflow/issues/9827

**Comment by mgerlach** [comment]
Hi, we migrated from 1.10.10 to 1.10.11 today. We are using Kubernetes Pod Operator and DO NOT specify ephemeral-storage in resources. Now our cluster (OpenShift 3.9) responds with HTTP 422
`Forbidden: ResourceEphemeralStorage field disabled by feature-gate for ResourceRequirements","field":"spec.containers[0].resources.limits"`

Looking at the JSON, this was sent to OpenShift bei the operator:
`
 "resources": {
          "limits": {
            "cpu": "4",
            "memory": "16Gi",
            "nvidia.com/gpu": null,
            "ephemeral-storage": null
          },
          "requests": {
            "cpu": "500m",
            "memory": "16Gi",
            "ephemeral-storage": null
          }
        },
`

I think it's an error to send the attribute at all if it is not specified. Any opinions on this?

**Comment by leonardoam** [comment]
Hi @mgerlach , it was in fact a bug introduced on 1.10.11 (see issues mentioned earlier). It will be fixed in 1.10.12 which is already under release process.

I believe the current workaround is specifying the missing resources in configs until next release.

**Comment by leonardoam** [comment]
New release 1.10.12 should be available on PyPI by now. It may solve your problem.

**Pull request merged**